pkg/filter/validator/basicauth.go

```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package validator
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/tg123/go-htpasswd"
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/megaease/easegress/pkg/cluster"
+	httpcontext "github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/supervisor"
+	"github.com/megaease/easegress/pkg/util/httpheader"
+	"github.com/megaease/easegress/pkg/util/yamltool"
+)
+
+type (
+	// EtcdSpec defines etcd prefix and which yaml entries are the username and password. For example spec
+	//   prefix: "/creds/"
+	//   usernameKey: "user"
+	//   passwordKey: "pw"
+	// expects the yaml to be stored with key /custom-data/creds/{id} in following yaml (extra keys are allowed)
+	//   user: doge
+	//   pw: {encrypted or plain text password}
+	EtcdSpec struct {
+		Prefix      string `yaml:"prefix" jsonschema:"onitempty"`
+		UsernameKey string `yaml:"usernameKey" jsonschema:"omitempty"`
+		PasswordKey string `yaml:"passwordKey" jsonschema:"omitempty"`
```

Comment:
Please also include default values for these two fields, as `Prefix`.

But I prefer the name of the two fields to be hard-coded, it seems a little over-flexible to make them configurable, maybe it is my fault that misunderstood your question and gave a wrong answer.

Reply:
It's true that hard coding them makes configuration simpler, I'll change that!